### ISSUE TITLE
Feature Request - Break on panic

### ISSUE THREAD

**Issue by GrumpyMetalGuy** [intellij-rust/intellij-rust#4763]
It would be extremely helpful to have the debugger break when a panic occurs. Seeing the stack trace and any relevant context at the point of panic would make debugging much easier.

**Comment by ortem** [comment]
You can add the following command to your [.gdbinit/.lldbinit](https://www.jetbrains.com/help/clion/configuring-debugger-options.html#gdbinit-lldbinit):
* `breakpoint set -n rust_panic` for LLDB
* `break rust_panic` for GDB

The execution will stop inside `panic` call
<img width="1312" alt="Screenshot 2020-02-05 at 12 33 01" src="https://user-images.githubusercontent.com/4854600/73829291-c7afff00-4813-11ea-9942-2634a74b5577.png">

Then you can select the desired frame and to look at local variables and so on
<img width="1312" alt="Screenshot 2020-02-05 at 12 33 08" src="https://user-images.githubusercontent.com/4854600/73829309-cd0d4980-4813-11ea-853f-0b6b67e254cb.png">


Does this solution work in your case? If so, we'll consider adding such a setting inside IDE.

**Comment by GrumpyMetalGuy** [comment]
Sorry for the delay in replying, just got around to trying this out now. Probably user error on my part, but not able to make it work. I'm on Windows, have installed MSys2 64 bit, setup the toolchain to point to it, and added a .gdbinit file in my C:\Users\<username> folder with the following contents:

set auto-load local-gdbinit on
add-auto-load-safe-path /
break rust_panic

I also tried adding a .gdbinit in the root of my project folder, adding the local settings as outlined at https://www.jetbrains.com/help/clion/configuring-debugger-options.html#gdbinit-lldbinit, but it still doesn't seem to stop.

I'm guessing I just don't have the files setup correctly, if anyone has an example .gdbinit I could shamelessly copy-pasta, that'd be awesome!

**Comment by GrumpyMetalGuy** [comment]
OK, made some progress. I set a breakpoint early in my program, and when I hit it, switched to the GDB view. One of the first things I see in the output window is:

No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) [answered N; input not from terminal]
No source file named C:/Development/Projects/rust/AdventOfCode2019/src/bin/day6.rs.
No source file named C:/Development/Projects/rust/AdventOfCode2019/src/intcode.rs.
No source file named C:/Development/Projects/rust/AdventOfCode2019/src/intcode.rs.
No source file named C:/Development/Projects/rust/AdventOfCode2019/src/bin/day11.rs.
**unset env HOME**

So for some reason, it's unsetting HOME, which I think may be making it lose the location of the .gdbinit file?

If I manually type it:

(gdb) break rust_panic
Breakpoint 6 at 0x42a6f3: file src\libstd/panicking.rs, line 521.

Thread 1 hit Breakpoint 6, rust_panic () at src\libstd/panicking.rs:521
521	        __rust_start_panic(obj as usize)

It works as intended.

So I guess I'm not sure why it's unsetting my HOME var? For reference, I'm using MSys2-64 bit, installed to C:\ProgramData\MSys2. 

Also, not sure how related this is, but when I typed this, got some kind of exception:

(gdb) show
ada print-signatures:  Whether the output of formal and return types for functions in the overloads selection menu is activated is on.
ada trust-PAD-over-XVS:  Whether an optimization trusting PAD types over XVS types is activated is on.
agent:  Debugger's willingness to use agent in inferior as a helper is off.
annotate:  Annotation_level is 0.
architecture:  The target architecture is set automatically (currently i386:x86-64)
<snip>
confirm:  Whether to confirm potentially dangerous operations is on.
conv:  $bpnum = 5
$_regex = <internal function _regex>
$_streq = <internal function _streq>
$_strlen = <internal function _strlen>
$_memeq = <internal function _memeq>
$_any_caller_matches = <internal function _any_caller_matches>
$_any_caller_is = <internal function _any_caller_is>
$_caller_matches = <internal function _caller_matches>
$_caller_is = <internal function _caller_is>
$_as_string = <internal function _as_string>
$_inferior = 1
$_isvoid = <internal function _isvoid>
$_sdata = ()
$_gthread = 1
$_thread = 1
$_probe_arg11 = <error: No probe at PC 0x000000000040262e>
$_probe_arg10 = <error: No probe at PC 0x000000000040262e>
$_probe_arg9 = <error: No probe at PC 0x000000000040262e>
$_probe_arg8 = <error: No probe at PC 0x000000000040262e>
$_probe_arg7 = <error: No probe at PC 0x000000000040262e>
$_probe_arg6 = <error: No probe at PC 0x000000000040262e>
$_probe_arg5 = <error: No probe at PC 0x000000000040262e>
$_probe_arg4 = <error: No probe at PC 0x000000000040262e>
$_probe_arg3 = <error: No probe at PC 0x000000000040262e>
$_probe_arg2 = <error: No probe at PC 0x000000000040262e>
$_probe_arg1 = <error: No probe at PC 0x000000000040262e>
$_probe_arg0 = <error: No probe at PC 0x000000000040262e>
$_probe_argc = <error: No probe at PC 0x000000000040262e>
$_siginfo = ()
$_exception = <error: not stopped at a C++ exception catchpoint>
Traceback (most recent call last):
  File "C:/Users/Peter Keyzers/AppData/Local/JetBrains/Toolbox/apps/CLion/ch-0/193.6494.38/bin/gdb/renderers\default\__init__.py", line 31, in lookup_type
    return gdb.parse_and_eval('(%s *) 0x0' % type_str).type.target()
gdb.error: syntax error in expression, near `) 0x0'.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:/Users/Peter Keyzers/AppData/Local/JetBrains/Toolbox/apps/CLion/ch-0/193.6494.38/bin/gdb/renderers\default\printers.py", line 135, in display_hint
    return '%s=%s' % (delegate_hint or '', self.to_string())
  File "C:/Users/Peter Keyzers/AppData/Local/JetBrains/Toolbox/apps/CLion/ch-0/193.6494.38/bin/gdb/renderers\default\printers.py", line 127, in to_string
    return '%s' % self.val.cast(gdb.lookup_type('void *'))
  File "C:/Users/Peter Keyzers/AppData/Local/JetBrains/Toolbox/apps/CLion/ch-0/193.6494.38/bin/gdb/renderers\default\__init__.py", line 33, in lookup_type
    raise e
  File "C:/Users/Peter Keyzers/AppData/Local/JetBrains/Toolbox/apps/CLion/ch-0/193.6494.38/bin/gdb/renderers\default\__init__.py", line 28, in lookup_type
    return lookup_type_orig(type_str)
gdb.error: No type named void *.
$_tlb = Traceback (most recent call last):
  File "C:/Users/Peter Keyzers/AppData/Local/JetBrains/Toolbox/apps/CLion/ch-0/193.6494.38/bin/gdb/renderers\default\__init__.py", line 31, in lookup_type
    return gdb.parse_and_eval('(%s *) 0x0' % type_str).type.target()
gdb.error: syntax error in expression, near `) 0x0'.

**Comment by GrumpyMetalGuy** [comment]
I've just had another go at this. I've manually set **HOME** in the debug configuration window, and I no longer see the message about HOME being unset. However, it still doesn't seem to be pulling in the .gdbinit file.

I've tried putting it in my user dir under C:\Windows, in the C:\ProgramData\MSys2 directory, in the root of my Rust project, and all don't seem to do anything. 

I can get it working by manually clicking the stack trace when it comes up and browsing to the source of the panic function in the rust source tree, then manually placing a breakpoint, but I'll have to do this for each project, so not ideal. If anyone has any suggestions for things I can try, please let me know!

**Comment by ortem** [comment]
@GrumpyMetalGuy Try to set `HOME` environment variable ([1](https://superuser.com/a/1183045)) and put `.gdbinit` into `HOME` directory ([2](https://stackoverflow.com/a/15399674)). Alternatively, if `.gdbinit` doesn't work anyway, you can set a breakpoint at the beginning of your program and execute `break rust_panic` command in CLion's GDB console when the debugger stops (it is a simulation of what `.gdbinit` actually does).

Regarding the exceptions unrelated to Rust, I'm afraid I cannot help you and could only recommend [CLion issue tracker](https://youtrack.jetbrains.com/issues/CPP).

**Comment by GrumpyMetalGuy** [comment]
@ortem Have tried those, but still no joy. Have tried with cygwin as well, but still no joy. I suspect I've got some environmental issue of some sort, but unless anyone else can chip in, I'm out of ideas :( 

I'd suggest implementing this as you had intended, and hopefully it will work for others. It definitely works manually! And who knows, maybe it'll work for me when you do it....

(BTW, still seeing this in the GDB window, so even after setting the env variable at the machine level, it's still unsetting it for some reason. This really seems like it could be causing the problem, but my Google-fu is failing me on this one and I can't find any other reports:

...
unset env HOSTNAME
unset env USER
unset env HOME
[New Thread 1840.0x2c34]
[New Thread 1840.0x23dc]

)

**Comment by GrumpyMetalGuy** [comment]
@ortem OK, I actually got this working :)  Needed one additional setting in the .gdbinit:

set breakpoint pending on
break rust_panic

With this in place, it breaks happily on panic. Happy days!

**Comment by kevleyski** [comment]
Is this a feature in the rust plugin? (IntelliJ and/or CLion)

**Comment by Kobzol** [comment]
@ortem It would be nice if these commands could be added automatically by the plugin. If you agree, I try to can send a PR.

**Comment by ortem** [comment]
@Kobzol Yes, that would be nice to have. I think you can try to implement it.

**Comment by GrumpyMetalGuy** [comment]
Thanks @Kobzol and @ortem !
